File: app/src/main/java/com/p1/uber/provides/GeofireProvider.java
package com.p1.uber.provides;

import com.firebase.geofire.GeoFire;
import com.firebase.geofire.GeoLocation;
import com.firebase.geofire.GeoQuery;
import com.google.android.gms.maps.model.LatLng;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;

public class GeofireProvider {

    private DatabaseReference mDatabase;
    private GeoFire mGeoFire;

    public GeofireProvider (String reference){
        mDatabase = FirebaseDatabase.getInstance().getReference().child(reference);
        mGeoFire =new GeoFire(mDatabase);
    }

    public  void saveLocation (String IdDriver, LatLng latlng){

        mGeoFire.setLocation(IdDriver, new GeoLocation(latlng.latitude, latlng.longitude));

    }

    public void removeLocation(String IdDriver){
        mGeoFire.removeLocation(IdDriver);
    }

    public GeoQuery getActiveDriver(LatLng latLng, double radius){
        GeoQuery geoQuery = mGeoFire.queryAtLocation(new GeoLocation(latLng.latitude,latLng.longitude),radius);
        geoQuery.removeAllListeners();
        return geoQuery;
    }

    public DatabaseReference getDriverLocation(String idDriver){
        return mDatabase.child(idDriver).child("l");
    }

    public DatabaseReference isDriverWorking(String idDriver){
        return FirebaseDatabase.getInstance().getReference().child("drivers_working").child(idDriver);
    }

}

File: app/src/main/java/com/p1/uberfares/provides/GeofireProvider.kt
package com.p1.uberfares.provides

import com.firebase.geofire.GeoFire
import com.firebase.geofire.GeoLocation
import com.firebase.geofire.GeoQuery
import com.google.android.gms.maps.model.LatLng
import com.google.firebase.database.DatabaseReference
import com.google.firebase.database.FirebaseDatabase

class GeofireProvider(reference: String?) {
    private val mDatabase: DatabaseReference
    private val mGeoFire: GeoFire

    init {
        mDatabase = FirebaseDatabase.getInstance().reference.child(reference!!)
        mGeoFire = GeoFire(mDatabase)
    }

    fun saveLocation(IdDriver: String?, latlng: LatLng) {
        mGeoFire.setLocation(IdDriver, GeoLocation(latlng.latitude, latlng.longitude))
    }

    fun removeLocation(IdDriver: String?) {
        mGeoFire.removeLocation(IdDriver)
    }

    fun getActiveDriver(latLng: LatLng, radius: Double): GeoQuery {
        val geoQuery =
            mGeoFire.queryAtLocation(GeoLocation(latLng.latitude, latLng.longitude), radius)
        geoQuery.removeAllListeners()
        return geoQuery
    }

    fun getActiveDriverExcept(latLng: LatLng, radius: Double): GeoQuery {
        val geoQuery =
            mGeoFire.queryAtLocation(GeoLocation(latLng.latitude, latLng.longitude), radius)
        geoQuery.removeAllListeners()
        return geoQuery
    }

    fun getDriverLocation(idDriver: String?): DatabaseReference {
        return mDatabase.child(idDriver!!).child("l")
    }



    fun isDriverWorking(idDriver: String?): DatabaseReference {
        return FirebaseDatabase.getInstance().reference.child("drivers_working").child(idDriver!!)
    }
}